src/main/scala/org/hammerlab/guacamole/windowing/SplitIterator.scala

```diff
@@ -0,0 +1,62 @@
+package org.hammerlab.guacamole.windowing
+
+object SplitIterator {
+  /**
+   * Split an iterator of key value pairs into separate iterators, one for each key, without buffering more data into
+   * memory than necessary. The keys must be integers between 0 and the provided upper bound.
+   *
+   * Given an iterator of (k, v) where 0 <= k < num and v is of arbitrary type, returns a sequence of length num. The
+   * i'th element in the sequence is an iterator over the values v where k = i.
+   *
+   * @param num number of keys. Each key k must be 0 <= k < num.
+   * @param iterator source iterator of (key, value) pairs. It will be consumed as the result iterators are read.
+   * @tparam V
+   * @return a buffered iterator for each key in 0 .. num.
+   */
+  def split[V](num: Int, iterator: Iterator[(Int, V)]): Seq[BufferedIterator[V]] = {
+    val buffers = (0 until num).map(_ => new collection.mutable.Queue[V])
+    (0 until num).map(index => new SplitIterator[V](index, buffers, iterator))
```

Comment:
Yeah, good call. IIRC we used to use `partition` in this way a while ago to split out each reads for each chromosome into a separate iterator, but the overhead of having to nest a `partition` for each key value was substantial (if there are N keys, each element can get buffered N times).